vqenc Buffer Overflow in KGL VQ Example

I've confirmed that this started happening once I bumped my host compiler to GCC-14, but when rebuilding with Clang, the issue goes away...

Anyway, building the KGL VQ example now fails with:
```
make -C  vq
make[4]: Entering directory '/opt/toolchains/dc/kos/examples/dreamcast/kgl/basic/vq'
rm -f vq-example.elf 
/opt/toolchains/dc/kos/utils/vqenc/vqenc -t -v fruit.jpg
encoding fruit.jpg.. *** buffer overflow detected ***: terminated
make[4]: *** [Makefile:23: fruit.vq] Aborted (core dumped)
make[4]: Leaving directory '/opt/toolchains/dc/kos/examples/dreamcast/kgl/basic/vq'
make[3]: *** [Makefile:16: _dir_vq] Error 2
make[3]: Leaving directory '/opt/toolchains/dc/kos/examples/dreamcast/kgl/basic'
make[2]: *** [Makefile:8: all] Error 2
make[2]: Leaving directory '/opt/toolchains/dc/kos/examples/dreamcast/kgl'
/bin/sh: 1: exit: Illegal number: -1
make[1]: *** [Makefile:16: all] Error 2
make[1]: Leaving directory '/opt/toolchains/dc/kos/examples/dreamcast'
make: *** [Makefile:8: all] Error 2
```

...which makes me think that probably there was always some buffer overflow issue, and only now GCC has gotten smart enough to detect it at runtime? I'll investigate when I have time, but hopefully someone beats me to it? ;)

I bet it'll be easy to find the issue with GDB.

--- a/utils/vqenc/vqenc.c
+++ b/utils/vqenc/vqenc.c
@@ -684,30 +684,21 @@ static void place_quads(context_t *cb, mipmap_t *m) {
 
 
 static const char *figure_outfilename(const char *f, const char *newext) {
-    char *newname;
     char *ext;
+    size_t namelen;
+    char *newname;
 
     ext = strrchr(f, '.');
 
-    if(ext == NULL) {
-        newname = (char *)malloc(strlen(f) + strlen(newext) + 2);
+    namelen = (ext == NULL) ? strlen(f) : (ext - f);
 
-        if(newname) {
-            sprintf(newname, "%s.%s", f, newext);
-        }
-    }
-    else {
-        int len;
+    newname = (char *)calloc(namelen + strlen(newext) + 2, sizeof(char));
 
-        len = (ext - f) + strlen(newext) + 2;
-        newname = (char *)malloc(len);
+    if(!newname) return NULL;
 
-        if(newname) {
-            strcpy(newname, f);
-            ext = strrchr(newname, '.') + 1;
-            strcpy(ext, newext);
-        }
-    }
+    strncpy(newname, f, namelen);
+    strcat(newname, ".");
+    strcat(newname, newext);
 
     return newname;
 }

